rust_src/src/files.rs

```diff
@@ -0,0 +1,111 @@
+extern crate libc;
+extern crate rand;
+
+use std::ffi::{CStr, CString};
+use std::io;
+#[cfg(unix)]
+use libc::{O_CLOEXEC, O_EXCL, O_RDWR, O_CREAT, open};
+
+#[cfg(test)]
+use std::{env, ptr};
+
+use self::rand::Rng;
+
+const NUM_RETRIES: usize = 50;
+
+//@TODO The third parameter of this will be an error code
+#[no_mangle]
+pub extern "C" fn rust_make_temp(template: *mut libc::c_char,
+                                 flags: libc::c_int,
+                                 _: *mut libc::c_int)
+                                 -> libc::c_int {
+
+    let rust_template = unsafe { CStr::from_ptr(template) };
+    let mut owned_template_copy = rust_template.to_string_lossy().into_owned();
+    let mut file_handle = -1;
+    for _ in 0..NUM_RETRIES {
+        generate_temporary_filename(&mut owned_template_copy);
+        let attempt_name = CString::new(owned_template_copy.clone()).unwrap();
+        file_handle = match open_temporary_file(&attempt_name, flags) {
+            Ok(file) => file,
+            Err(ref e) if e.kind() == io::ErrorKind::AlreadyExists => continue,
+            Err(_) => -1
+        };
+
+        if file_handle != -1 {
+            unsafe { libc::strcpy(template, attempt_name.as_ptr()); }           
+            break;
+        }
+    };
+
+    file_handle
+}
+
+fn generate_temporary_filename(name: &mut String) {
```

Comment:
I chose to make this take a mutable reference, as https://doc.rust-lang.org/book/references-and-borrowing.html calls out taking ownership and then returning as 'unidiomatic'. I am opening to changing it if we prefer the style you described.